### PR TITLE
Hotfix ETP-2292: Fix session problem on Create Lines from manual process

### PR DESCRIPTION
**Pull request by Luciano-Palacio**

### CODE REVIEW

File: src/com/etendoerp/metadata/utils/LegacyUtils.java [hunk 1]
```diff
@@ -96,4 +101,22 @@ public static boolean isLegacyPath(String path) {
...
+     * Finds the Active Tab ID given a Window ID and Table ID.
+     */
+    public static String findTabIdByWindowAndTable(String windowId, String tableId) {
+        if (StringUtils.isEmpty(windowId) || StringUtils.isEmpty(tableId)) {
+            return null;
+        }
+        OBQuery<Tab> query = OBDal.getInstance().createQuery(
+            Tab.class,
+            "where window.id = :windowId and table.id = :tableId and active = true order by sequenceNumber"
+        );
```
Comment (etendobot) [inline src/com/etendoerp/metadata/utils/LegacyUtils.java:122]: **Suggestion**

The query logic is sound. Ensure that the 'active' property is consistent with the AD_Tab entity definition (usually 'active' in DAL).

```java
"where window.id = :windowId and table.id = :tableId and active = true order by sequenceNumber"
```

**Rationale**: The query is correct and efficient. Using setMaxResult(1) ensures only the necessary record is retrieved.

File: src/com/etendoerp/metadata/utils/LegacyUtils.java [hunk 2]
```diff
@@ -96,4 +101,22 @@ public static boolean isLegacyPath(String path) {
...
+        query.setNamedParameter("windowId", windowId);
+        query.setNamedParameter("tableId", tableId);
+        query.setMaxResult(1);
+        List<Tab> tabs = query.list();
+        return tabs.isEmpty() ? null : tabs.get(0).getId();
+    }
```
Comment (etendobot) [inline src/com/etendoerp/metadata/utils/LegacyUtils.java:128]: **Suggestion**

Consider using OBContext.setAdminMode() within this utility method if it might be called from contexts where the current user does not have direct access to the AD_Tab table metadata.

```java
return tabs.isEmpty() ? null : tabs.get(0).getId();
```

**Rationale**: Retrieving only the ID from the first result of a limited query is a common and safe pattern in Etendo.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java [hunk 3]
```diff
@@ -865,4 +869,24 @@ private String injectCodeAfterFunctionCall(String originalRes, String originalFu
...
     }
+
+    private static void handleCreateFromSession(HttpServletRequest req, String path, HttpSession session) {
+        if (LegacyPaths.CREATE_FROM_HTML.equals(path) && "SAVE".equals(req.getParameter("Command"))) {
+            String windowId = req.getParameter("inpWindowId");
+            String tableId = req.getParameter("inpTableId");
+            String sessionKey = "CREATEFROM|TABID";
+
+            if (!LegacyUtils.isMutableSessionAttribute(sessionKey)) {
+                throw new InternalServerException("Attempt to set forbidden session key: " + sessionKey);
```
Comment (etendobot) [inline src/com/etendoerp/metadata/http/LegacyProcessServlet.java:880]: > [!WARNING]
**⚠️ Blocking Issue**

You must update LegacyUtils.java to include 'CREATEFROM|TABID' in the MUTABLE_SESSION_ATTRIBUTES set.

```java
// Ensure the key is added to LegacyUtils.MUTABLE_SESSION_ATTRIBUTES
if (!LegacyUtils.isMutableSessionAttribute(sessionKey)) {
```

**Rationale**: The sessionKey 'CREATEFROM|TABID' is not part of the MUTABLE_SESSION_ATTRIBUTES set in LegacyUtils.java. This will cause an InternalServerException to be thrown every time this condition is met, breaking the intended functionality.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java [hunk 4]
```diff
@@ -865,4 +869,24 @@ private String injectCodeAfterFunctionCall(String originalRes, String originalFu
         matcher.appendTail(result);
         return result.toString();
     }
+
+    private static void handleCreateFromSession(HttpServletRequest req, String path, HttpSession session) {
+        if (LegacyPaths.CREATE_FROM_HTML.equals(path) && "SAVE".equals(req.getParameter("Command"))) {
+            String windowId = req.getParameter("inpWindowId");
```
Comment (etendobot) [inline src/com/etendoerp/metadata/http/LegacyProcessServlet.java:875]: > [!WARNING]
**⚠️ Blocking Issue**

Verify that LegacyPaths.java is updated to include the CREATE_FROM_HTML constant.

```java
// Ensure CREATE_FROM_HTML is defined in LegacyPaths.java
```

**Rationale**: The constant LegacyPaths.CREATE_FROM_HTML does not appear to be defined in the LegacyPaths utility class, which will result in a compilation error.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java [hunk 5]
```diff
@@ -865,4 +869,24 @@ private String injectCodeAfterFunctionCall(String originalRes, String originalFu
...
+            String tableId = req.getParameter("inpTableId");
+            String sessionKey = "CREATEFROM|TABID";
+
+            if (!LegacyUtils.isMutableSessionAttribute(sessionKey)) {
+                throw new InternalServerException("Attempt to set forbidden session key: " + sessionKey);
+            }
+
+            String tabId = LegacyUtils.findTabIdByWindowAndTable(windowId, tableId);
+            if (tabId != null) {
```
Comment (etendobot) [inline src/com/etendoerp/metadata/http/LegacyProcessServlet.java:884]: > [!WARNING]
**⚠️ Blocking Issue**

Ensure that the findTabIdByWindowAndTable method is implemented and exported in LegacyUtils.java.

```java
// Ensure findTabIdByWindowAndTable is implemented in LegacyUtils.java
```

**Rationale**: The method LegacyUtils.findTabIdByWindowAndTable does not appear to exist in the LegacyUtils utility class, which will cause a compilation failure.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java [hunk 6]
```diff
@@ -522,7 +526,7 @@ private void writeFinalResponse(HttpServletResponse res,
     /**
      * Creates a new session in the database (AD_Session table).
      * This method is a duplication of the logic found in
-     * {@link org.openbravo.authentication.AuthenticationManager#createDBSession(HttpServletRequest, String, String, String)}.
+     * {@link org.openbravo.authentication.AuthenticationManager# createDBSession(HttpServletRequest, String, String, String)}.
      *
```
Comment (etendobot) [inline src/com/etendoerp/metadata/http/LegacyProcessServlet.java:530]: **Suggestion**

Remove the space between the '#' and the method name.

```java
* {@link org.openbravo.authentication.AuthenticationManager#createDBSession(HttpServletRequest, String, String, String)}.
```

**Rationale**: The addition of a space after the '#' character breaks the Javadoc link syntax. The Class#method pattern must be preserved without spaces.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java [hunk 7]
```diff
@@ -327,6 +330,7 @@ private void processLegacyRequest(HttpServletRequest req, HttpServletResponse re
                 authenticateWithToken(wrappedRequest, token);
             }
             preprocessRequest(req, wrappedRequest);
+            handleCreateFromSession(req, path, req.getSession(true));
```
Comment (etendobot) [inline src/com/etendoerp/metadata/http/LegacyProcessServlet.java:334]: **Suggestion**

Pass only the request to handleCreateFromSession and create the session inside the method only when the specific path and command conditions are met.

```java
handleCreateFromSession(req, path);

// Inside handleCreateFromSession:
if (LegacyPaths.CREATE_FROM_HTML.equals(path) && "SAVE".equals(req.getParameter("Command"))) {
    HttpSession session = req.getSession(true);
    // ... logic
}
```

**Rationale**: Calling req.getSession(true) creates a session if one doesn't exist. Doing this for every legacy request before checking if the path matches CREATE_FROM_HTML can lead to unnecessary session creation and resource consumption.

File: src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java [hunk 8]
```diff
@@ -60,6 +65,7 @@ public class LegacyProcessServletTest extends OBBaseTest {
     public static final String REDIRECT = "/redirect";
     public static final String LOCATION = "location";
     private static final String USER_ID_KEY = "userId";
+    private static final String CREATE_FROM_SESSION_KEY = "CREATEFROM|TABID";
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java:72]: **Suggestion**

Consider defining this constant in LegacyProcessServlet as public or in a dedicated constants class to ensure consistency between the code and the test.

```java
// Ensure this constant matches the one used in LegacyProcessServlet or move it to a shared constants class.
```

**Rationale**: Using a local private constant for a session key that is shared with the production code can lead to mismatches if the key is changed in one place but not the other. Ideally, this key should be defined in a shared constants class or in the LegacyProcessServlet itself and accessed from the test.

File: src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java [hunk 9]
```diff
@@ -11,6 +11,12 @@
 import org.mockito.junit.jupiter.MockitoExtension;
 import org.openbravo.base.provider.OBProvider;
 import org.openbravo.model.ad.ui.Process;
+import org.openbravo.dal.service.OBDal;
+import org.openbravo.dal.service.OBQuery;
+import org.openbravo.model.ad.ui.Tab;
+import java.util.List;
+import static org.mockito.ArgumentMatchers.eq;
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java:18]: > [!WARNING]
**⚠️ Blocking Issue**

Include the missing Mockito imports for static and dynamic mocking.

```java
import java.util.List;
import org.mockito.MockedStatic;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
```

**Rationale**: The test code utilizes Mockito's static mocking features (MockedStatic, mockStatic) and standard mocking methods (mock, when, verify). These require explicit imports to compile correctly. Their absence in the added lines suggests a build failure will occur.

File: src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java [hunk 10]
```diff
@@ -20,6 +26,8 @@
 class LegacyUtilsTest {
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java:27]: **Suggestion**

Add a unit test to verify that the LegacyUtils utility class cannot be instantiated.

```java
class LegacyUtilsTest {

  @Test
  void testConstructorIsPrivate() throws NoSuchMethodException {
    Constructor<LegacyUtils> constructor = LegacyUtils.class.getDeclaredConstructor();
    assertTrue(Modifier.isPrivate(constructor.getModifiers()), "Constructor should be private");
    constructor.setAccessible(true);
    InvocationTargetException exception = assertThrows(InvocationTargetException.class, constructor::newInstance);
    assertTrue(exception.getCause() instanceof UnsupportedOperationException);
  }
```

**Rationale**: It is a standard Etendo development pattern to include a test verifying that utility classes are final and have a private constructor to prevent instantiation.

File: src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java [hunk 11]
```diff
@@ -174,4 +182,66 @@ void staticSetsAreInitializedCorrectly() throws Exception {
...
+
+  @Test
+  void findTabIdByWindowAndTableReturnsNullForNullTableId() {
+    assertNull(LegacyUtils.findTabIdByWindowAndTable(WINDOW_ID, null));
+  }
+
+  @Test
+  void findTabIdByWindowAndTableReturnsNullForEmptyWindowId() {
+    assertNull(LegacyUtils.findTabIdByWindowAndTable("", TABLE_ID));
+  }
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java:204]: **Suggestion**

Include additional edge cases for empty string parameters in the findTabIdByWindowAndTable test suite.

```java
assertNull(LegacyUtils.findTabIdByWindowAndTable("", TABLE_ID));
    assertNull(LegacyUtils.findTabIdByWindowAndTable(WINDOW_ID, ""));
    assertNull(LegacyUtils.findTabIdByWindowAndTable("", ""));
```

**Rationale**: Expanding the test cases to include empty table IDs and both parameters being empty ensures the method handles all edge cases of invalid string input consistently.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java (hunk 3 above)
Comment (Luciano-Palacio) [reply src/com/etendoerp/metadata/http/LegacyProcessServlet.java:880]: Done! :heavy_check_mark:

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java (hunk 4 above)
Comment (Luciano-Palacio) [reply src/com/etendoerp/metadata/http/LegacyProcessServlet.java:875]: Done! :heavy_check_mark:

File: src/com/etendoerp/metadata/http/LegacyProcessServlet.java (hunk 5 above)
Comment (Luciano-Palacio) [reply src/com/etendoerp/metadata/http/LegacyProcessServlet.java:884]: Done! :heavy_check_mark:

File: src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java (hunk 9 above)
Comment (Luciano-Palacio) [reply src-test/src/com/etendoerp/metadata/utils/LegacyUtilsTest.java:18]: Done! :heavy_check_mark:

File: src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java [hunk 12]
```diff
@@ -60,6 +65,10 @@ public class LegacyProcessServletTest extends OBBaseTest {
...
+    private static final String COMMAND_KEY = "Command";
+    private static final String INP_WINDOW_ID_KEY = "inpWindowId";
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java:70]: **Suggestion**

Consider renaming the constant to PARAM_COMMAND to maintain consistency with the existing naming convention for parameter-related constants in this file.

```java
private static final String PARAM_COMMAND = "Command";
```

**Rationale**: Following the established naming pattern in the file (e.g., PARAM_INP_KEY, PARAM_INP_WINDOW_ID) improves maintainability and readability.

File: src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java [hunk 13]
```diff
@@ -60,6 +65,10 @@ public class LegacyProcessServletTest extends OBBaseTest {
...
+    private static final String INP_TABLE_ID_KEY = "inpTableId";
```
Comment (etendobot) [inline src-test/src/com/etendoerp/metadata/http/LegacyProcessServletTest.java:71]: **Suggestion**

Verify if 'PARAM_INP_WINDOW_ID' already exists in this class and consider reusing it. If the casing 'inpWindowId' (capital W) is specifically required for this test, consider renaming this constant to 'PARAM_INP_WINDOW_ID' to match the existing pattern.

```java
private static final String PARAM_INP_WINDOW_ID = "inpWindowId";
```

**Rationale**: Consistency with existing 'PARAM_...' naming and potential duplication check. There is an existing 'PARAM_INP_WINDOW_ID' constant in the KB version of this file; verify if it can be reused or if the specific casing 'inpWindowId' is required.

**Review by etendobot — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

File: src/com/etendoerp/metadata/utils/LegacyUtils.java [hunk 14]
```diff
@@ -96,4 +101,29 @@ public static boolean isLegacyPath(String path) {
...
+
+    /**
+     * Finds the ID of the first active tab that belongs to the given window and table.
+     * The result is ordered by sequence number to ensure deterministic output when
+     * multiple tabs reference the same window and table combination.
+     *
+     * @param windowId the ID of the AD_Window to filter by
+     * @param tableId  the ID of the AD_Table to filter by
+     * @return the ID of the matching tab, or {@code null} if either parameter is blank
+     *         or no active tab is found for the given window and table
```
Comment (etendobot) [inline src/com/etendoerp/metadata/utils/LegacyUtils.java:113]: **Suggestion**

Update the Javadoc to use 'empty' instead of 'blank' to match the `StringUtils.isEmpty()` check in the implementation.

```java
* @return the ID of the matching tab, or {@code null} if either parameter is empty
```

**Rationale**: The implementation in the context lines uses `StringUtils.isEmpty()`, which does not consider whitespace-only strings as empty. Using the word 'blank' in the Javadoc suggests the use of `StringUtils.isBlank()`. To be technically accurate, 'empty' should be used if the logic remains `isEmpty()`.

**Review by etendobot — APPROVED** [review]

**Review by Emi-Polliotti — APPROVED** [review]
